app/src/components/api.tsx

```diff
@@ -168,23 +168,38 @@ export function setupSelection(
   }, [selectedIds]);
 }
 
-export function setupStep(token: string, setStep: any, step: any) {
-  const conInterfaceRef = useRef<any>(undefined);
+export function setupStep(
+  token: string,
+  setStep: (items: any) => void,
+  step: any,
+) {
+  // Keep conInterfaceRef stable across renders
+  const conInterfaceRef = useRef<any>(null);
   const updateByRefreshRef = useRef(false);
 
-  useEffect(() => {
-    const con = new znsocket.Dict({
+  // Memoize `con` to keep it consistent
+  const con = useMemo(() => {
+    return new znsocket.Dict({
       client: client,
       key: "room:" + token + ":step",
     });
+  }, [token]);
```

Comment:
Is it a good idea to memoize this, when on/off is called in useEffect?

Reply:
With https://socket.io/how-to/use-with-react it might be best to define all the znsocket objects in a module and import